### PR TITLE
added cpp solution to usaco-717

### PR DESCRIPTION
**Pull request by RedBlueBird**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/usaco-717.mdx [hunk 1]
```diff
@@ -2,14 +2,83 @@
...
 
 <LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+typedef pair<int,int> pii;
+
+//Possible directions a cow can go to eat grass
```
Comment (ryanchou-dev) [inline solutions/usaco-717.mdx:20]:
```suggestion
// Possible directions a cow can go to eat grass
```

File: solutions/usaco-717.mdx [hunk 2]
```diff
@@ -2,14 +2,83 @@
...
+        int x, y;
+        tie(x, y) = dijkstra.top().ss;
+        dijkstra.pop();
+        if (visited[x][y]){
+			continue;
+		}
+        visited[x][y] = true;
+        for (int i = 0; i < 16; i++){
+            int dx = x + Dx[i], dy = y + Dy[i];
+            if (dx < 0 || dx >= n || dy < 0 || dy >= n) continue;
```
Comment (ryanchou-dev) [inline solutions/usaco-717.mdx:60]: braces

**Review by ryanchou-dev — COMMENTED** [review]
looks good

File: solutions/usaco-717.mdx [hunk 3]
```diff
@@ -2,14 +2,83 @@
...
+        }
+    }
+
+    /* 
+     * There are 16 possible tiles a cow can reach in 3 moves
+     * Form edges between those tiles and with assigned weight:
+     * (the time it takes to eat the tile) + 3 * t
+     * If the tile is less than 3 manhattan distance to farmer John's tile,
+     * form an edge between those two tiles also.
+     * The problem is now simple dijkstra
```
Comment (ryanchou-dev) [inline solutions/usaco-717.mdx:43]:
```suggestion
     * The problem can now be solved with Dijkstra's
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
indent with tabs too

File: solutions/usaco-717.mdx [hunk 4]
```diff
@@ -2,14 +2,84 @@
...
+    while (dijkstra.size()){
+        int x, y;
+        tie(x, y) = dijkstra.top().ss;
+        dijkstra.pop();
+        if (visited[x][y]){
+			continue;
+		}
+        visited[x][y] = true;
+        for (int i = 0; i < 16; i++){
+            int dx = x + Dx[i], dy = y + Dy[i];
```
Comment (envyaims) [inline solutions/usaco-717.mdx:60]: just something minor:
`dx` typically stands for delta x, which means change in x. However `dx` in this case is the actual new x. I would recommend using something like `newx` here and decapitalize the array variable because it can get confusing

File: solutions/usaco-717.mdx [hunk 5]
```diff
@@ -2,14 +2,84 @@
...
+     * The problem is now simple dijkstra
+     */
+    vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
+    vector<vector<bool>> visited(n, vector<bool>(n));
+    dist[0][0] = 0;
+    priority_queue<pair<int, pii>> dijkstra;
+    dijkstra.push(make_pair(0, make_pair(0,0)));
+    while (dijkstra.size()){
+        int x, y;
+        tie(x, y) = dijkstra.top().ss;
```
Comment (envyaims) [inline solutions/usaco-717.mdx:53]: you dont have `.ss` as a macro here

File: solutions/usaco-717.mdx [hunk 6]
```diff
@@ -2,14 +2,84 @@
...
+     * If the tile is less than 3 manhattan distance to farmer John's tile,
+     * form an edge between those two tiles also.
+     * The problem is now simple dijkstra
+     */
+    vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
+    vector<vector<bool>> visited(n, vector<bool>(n));
+    dist[0][0] = 0;
+    priority_queue<pair<int, pii>> dijkstra;
+    dijkstra.push(make_pair(0, make_pair(0,0)));
+    while (dijkstra.size()){
```
Comment (envyaims) [inline solutions/usaco-717.mdx:51]: optional: `!dijkstra.empty()` it the common way to do this and would probably be more intuitive for readers, but its up to you

File: solutions/usaco-717.mdx [hunk 7]
```diff
@@ -2,14 +2,84 @@
...
+     * Form edges between those tiles and with assigned weight:
+     * (the time it takes to eat the tile) + 3 * t
+     * If the tile is less than 3 manhattan distance to farmer John's tile,
+     * form an edge between those two tiles also.
+     * The problem is now simple dijkstra
+     */
+    vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
+    vector<vector<bool>> visited(n, vector<bool>(n));
+    dist[0][0] = 0;
+    priority_queue<pair<int, pii>> dijkstra;
```
Comment (envyaims) [inline solutions/usaco-717.mdx:49]: also optional: instead of a pair macro with nested pairs just use `array<int,3>`
Also you should clarify that the pair goes from `{distance, {row, col}}`

**Review by envyaims — CHANGES_REQUESTED** [review]

File: solutions/usaco-717.mdx [hunk 8]
```diff
@@ -2,14 +2,84 @@
...
+    int n, t;
+    fin >> n >> t;
+    vector<vector<int>> field(n, vector<int>(n));
+    for (int i = 0; i < n; i++){
+        for (int j = 0; j < n; j++){
+            fin >> field[i][j];
+        }
+    }
+
+    /*
```
Comment (envyaims) [inline solutions/usaco-717.mdx:38]: instead of having this huge comment maybe put it outside the code?

File: solutions/usaco-717.mdx (hunk 4 above)
Comment (RedBlueBird) [reply solutions/usaco-717.mdx]: Ok good idea, I will address that

File: solutions/usaco-717.mdx [hunk 9]
```diff
@@ -2,14 +2,93 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+typedef pair<int,int> pii;
+#define ff first
+#define ss second
+
+// Possible directions a cow can go to eat grass
+int Dx[] = {1, 0, -1, 0, 3, 0, -3, 0, 2, 2, 1, 1, -1, -1, -2, -2};
```
Comment (envyaims) [inline solutions/usaco-717.mdx:36]: decapitalize

**Review by envyaims — APPROVED** [review]
seems fine

File: solutions/usaco-717.mdx [hunk 10]
```diff
@@ -2,14 +2,93 @@
...
+		for (int j = 0; j < n; j++){
+			fin >> field[i][j];
+		}
+	}
+
+	vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
+	vector<vector<bool>> visited(n, vector<bool>(n));
+	dist[0][0] = 0;
+	priority_queue<pair<int, pii>> dijkstra; // {distance, {row, col}}
+	dijkstra.push(make_pair(0, make_pair(0,0)));
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:56]: Add comments for this.

File: solutions/usaco-717.mdx [hunk 11]
```diff
@@ -2,14 +2,93 @@
...
+		for (int i = 0; i < 16; i++){
+			int newx = x + Dx[i], newy = y + Dy[i];
+			if (newx < 0 || newx >= n || newy < 0 || newy >= n){
+				continue;
+			}
+			if (dist[newx][newy] > dist[x][y] + 3 * t + field[newx][newy]){
+				dist[newx][newy] = dist[x][y] + 3 * t + field[newx][newy];
+				dijkstra.push(make_pair(-dist[newx][newy], 
+							  make_pair(newx,newy)));
+			}
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:74]: Add Comments for this

File: solutions/usaco-717.mdx [hunk 12]
```diff
@@ -2,14 +2,93 @@
...
+	vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
+	vector<vector<bool>> visited(n, vector<bool>(n));
+	dist[0][0] = 0;
+	priority_queue<pair<int, pii>> dijkstra; // {distance, {row, col}}
+	dijkstra.push(make_pair(0, make_pair(0,0)));
+	while (!dijkstra.empty()){
+		int x, y;
+		tie(x, y) = dijkstra.top().ss;
+		dijkstra.pop();
+		if (visited[x][y]){
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:61]:
```suggestion
		
		// check if this coordinate has already been visited
		if (visited[x][y]){
```

File: solutions/usaco-717.mdx [hunk 13]
```diff
@@ -2,14 +2,93 @@
...
 
+## Solution
+
+Instead of treating this problem with grids, let us consider from a more graphical perspective with nodes and edges.
+
+We first observe that because Bessie only feasts on grass tiles every 3 moves and the time to travel between adjacent tiles is a constant $T$. The two intermediary tiles Bessie used to travel does not actually matter. So we want to form a direct edge between each tile that are exactly 3 manhattan distances apart, with an weight of 
+
+$\{\text{The time to feast the target grass tile}\} + 3 * T$
+
+However, if the tile Bessie just feasted on is less than 3 manhattan distances away from Farmer John's tile, then it would be more desirable for Bessie to travel to Farmer John directly than take another 3-distanced tile. So let us too form a direct edge between those two tiles with an weight of
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:18]:
```suggestion
However, if the tile Bessie just feasted on is less than 3 units (using manhattan distance) away from Farmer John's tile, then it would be more desirable for Bessie to travel to Farmer John directly than take another 3-distanced tile. So, we also form a direct edge between those two tiles with a weight of 
```

File: solutions/usaco-717.mdx [hunk 14]
```diff
@@ -2,14 +2,93 @@
...
 title: Why Did the Cow Cross the Road
-author: Qi Wang
+author: Qi Wang, Jeffrey Zhang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_visitfj_gold_feb17.html)
 
+## Solution
+
+Instead of treating this problem with grids, let us consider from a more graphical perspective with nodes and edges.
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:12]:
```suggestion
Instead of treating this as a grid problem, let us try to approach it as a graph problem with nodes and edges.```

File: solutions/usaco-717.mdx [hunk 15]
```diff
@@ -2,14 +2,93 @@
...
+
+We first observe that because Bessie only feasts on grass tiles every 3 moves and the time to travel between adjacent tiles is a constant $T$. The two intermediary tiles Bessie used to travel does not actually matter. So we want to form a direct edge between each tile that are exactly 3 manhattan distances apart, with an weight of
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:14]:
```suggestion
We first observe that the two intermediary tiles Bessie used to travel do not actually matter because Bessie only feasts on grass tiles every 3 moves and the time to travel between adjacent tiles is a constant $T$. So we want to form a direct edge between each tile that is exactly 3 manhattan distances apart, with a weight of 
```

File: solutions/usaco-717.mdx (hunk 9 above)
Comment (Aadit-Ambadkar) [reply solutions/usaco-717.mdx:36]:
```suggestion
int dx[] = {1, 0, -1, 0, 3, 0, -3, 0, 2, 2, 1, 1, -1, -1, -2, -2};
```

File: solutions/usaco-717.mdx [hunk 16]
```diff
@@ -2,14 +2,93 @@
...
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+typedef pair<int,int> pii;
+#define ff first
+#define ss second
+
+// Possible directions a cow can go to eat grass
+int Dx[] = {1, 0, -1, 0, 3, 0, -3, 0, 2, 2, 1, 1, -1, -1, -2, -2};
+int Dy[] = {0, 1, 0, -1, 0, 3, 0, -3, 1, -1, 2, -2, 2, -2, 1, -1};
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:37]:
```suggestion
int dy[] = {0, 1, 0, -1, 0, 3, 0, -3, 1, -1, 2, -2, 2, -2, 1, -1};
```

File: solutions/usaco-717.mdx [hunk 17]
```diff
@@ -2,14 +2,93 @@
...
+	while (!dijkstra.empty()){
+		int x, y;
+		tie(x, y) = dijkstra.top().ss;
+		dijkstra.pop();
+		if (visited[x][y]){
+			continue;
+		}
+		visited[x][y] = true;
+		for (int i = 0; i < 16; i++){
+			int newx = x + Dx[i], newy = y + Dy[i];
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-717.mdx:66]:
```suggestion
			int newx = x + dx[i], newy = y + dy[i];
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]
Tl;dr make sure your grammar is correct (I would use Grammarly for this) and add more comments.

File: solutions/usaco-717.mdx [hunk 18]
```diff
@@ -2,14 +2,100 @@
...
+		int manhattan = n - x - 1 + n - y - 1;
+		if (manhattan < 3){
+			dist[n - 1][n - 1] = min(dist[n - 1][n - 1], 
+									 dist[x][y] + manhattan * t);
+		}
+	}
+
+	fout << dist[n - 1][n - 1];
+
+	return 0;
```
Comment (ryanchou-dev) [inline solutions/usaco-717.mdx:92]:
```suggestion
```

File: solutions/usaco-717.mdx [hunk 19]
```diff
@@ -2,14 +2,100 @@
...
+			}
+			// updating the shortest distance to this new location
+			if (dist[newx][newy] > dist[x][y] + 3 * t + field[newx][newy]){
+				dist[newx][newy] = dist[x][y] + 3 * t + field[newx][newy];
+				dijkstra.push(make_pair(-dist[newx][newy], 
+							  make_pair(newx,newy)));
+			}
+		}
+		int manhattan = n - x - 1 + n - y - 1;
+		if (manhattan < 3){
```
Comment (ryanchou-dev) [inline solutions/usaco-717.mdx:84]:
```suggestion
		if (manhattan < 3) {
```
space out braces like this in the rest of the code as well

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
just this 🎄

File: solutions/usaco-717.mdx [hunk 20]
```diff
@@ -2,14 +2,98 @@
...
 title: Why Did the Cow Cross the Road
-author: Qi Wang
+author: Qi Wang, Jeffrey Zhang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_visitfj_gold_feb17.html)
 
+## Solution
+
+Instead of treating this as a grid problem, let us try to approach it as a graph problem with nodes and edges.```
```
Comment (SansPapyrus683) [inline solutions/usaco-717.mdx:12]: what's with the 3 backticks at the end?

File: solutions/usaco-717.mdx [hunk 21]
```diff
@@ -2,14 +2,98 @@
...
+			}
+		}
+		int manhattan = n - x - 1 + n - y - 1;
+		if (manhattan < 3) {
+			dist[n - 1][n - 1] = min(dist[n - 1][n - 1], 
+									 dist[x][y] + manhattan * t);
+		}
+	}
+
+	fout << dist[n - 1][n - 1];
```
Comment (SansPapyrus683) [inline solutions/usaco-717.mdx:90]:
```suggestion
	ofstream("visitfj.out") << dist[n - 1][n - 1] << endl;
```

File: solutions/usaco-717.mdx [hunk 22]
```diff
@@ -2,14 +2,98 @@
...
+#define ff first
+#define ss second
+
+// Possible directions a cow can go to eat grass
+int dx[] = {1, 0, -1, 0, 3, 0, -3, 0, 2, 2, 1, 1, -1, -1, -2, -2};
+int dy[] = {0, 1, 0, -1, 0, 3, 0, -3, 1, -1, 2, -2, 2, -2, 1, -1};
+
+int main() {
+	ifstream fin("visitfj.in");
+	ofstream fout("visitfj.out");
```
Comment (SansPapyrus683) [inline solutions/usaco-717.mdx:41]:
```suggestion
```

File: solutions/usaco-717.mdx [hunk 23]
```diff
@@ -2,14 +2,98 @@
...
 <LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+typedef pair<int, int> pii;
+#define ff first
+#define ss second
```
Comment (SansPapyrus683) [inline solutions/usaco-717.mdx:33]: remove unused macros
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
yeah p good jeff

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
will merge after deployment
